### PR TITLE
Transfer alias mappings when joining an upgraded room

### PR DESCRIPTION
**Pull request by anoadragon453**
Fixes https://github.com/matrix-org/synapse/issues/6784

~~Needs some tests.~~ Sytest PR: https://github.com/matrix-org/sytest/pull/844

### CODE REVIEW

**Comment by clokep** [comment]
@anoadragon453 I forget exactly where we landed on this, but I believe this is still necessary after the other alias changes that were made? Does that sound accurate?

**Comment by anoadragon453** [comment]
> @anoadragon453 I forget exactly where we landed on this, but I believe this is still necessary after the other alias changes that were made? Does that sound accurate?

Yes, that's correct.

File: synapse/storage/data_stores/main/directory.py [hunk 1]
```diff
@@ -159,10 +160,28 @@ def _delete_room_alias_txn(self, txn, room_alias):
...
-    def update_aliases_for_room(self, old_room_id, new_room_id, creator):
+    def update_aliases_for_room(
+        self, old_room_id: str, new_room_id: str, creator: Optional[str] = None,
+    ):
+        """Repoint the aliases from one room to another
+
+        Args:
+            old_room_id:
+            new_room_id:
+            creator: The creator of the alias mapping. If None, don't update
```
Comment (richvdh) [inline synapse/storage/data_stores/main/directory.py]: this is unclear: I'm not sure if I should be setting it to the previous creator of the old alias, or the new one. And where it says "don't update", I initially read that as meaning that the alias wouldn't be updated at all.
```suggestion
            creator: The user to record as the creator of the new mapping. 
                If None, the creator will be left unchanged.
```

File: synapse/storage/data_stores/main/directory.py [hunk 2]
```diff
@@ -159,10 +160,28 @@ def _delete_room_alias_txn(self, txn, room_alias):
 
         return room_id
 
-    def update_aliases_for_room(self, old_room_id, new_room_id, creator):
+    def update_aliases_for_room(
+        self, old_room_id: str, new_room_id: str, creator: Optional[str] = None,
+    ):
+        """Repoint the aliases from one room to another
```
Comment (richvdh) [inline synapse/storage/data_stores/main/directory.py]:
```suggestion
        """Repoint all of the aliases for a given room, to a different room.
```

**Pull request merged**